drone/src/agent/executor.rs

```diff
@@ -265,30 +265,36 @@ impl<E: Engine> Executor<E> {
             }
 
             let next_state = loop {
-                //we ignore external state changes in a terminal state
-                //to avoid multiple destructor calls for backend.
-                state.terminal().then(|| recv.close());
-                tokio::select! {
-                    next_state = self.step(spawn_request, state) => break next_state,
-                    sig = recv.recv() => match sig {
-                        Some(Signal::Interrupt) => {
-                            tracing::info!("State may have updated externally.");
-                            continue;
-                        },
-                        Some(Signal::Terminate) => {
-                            break Ok(Some(BackendState::Terminated))
-                        },
-                        None => {
-                            if state.terminal() {
-                                tracing::info!(%state, "Channel closed, backend in terminal state");
-                            } else {
-                                tracing::error!(
-                                    %state, "Channel closed but backend not in terminal state");
+                if state.terminal() {
+                    //we stop the signal handler to ensure that there are no double terminations
+                    recv.close();
```

Comment:
Can you remind me why this is necessary? It seems like if the state is terminal, `self.step` will return `Ok(None)` and then we get to the end and `recv` will be dropped when it is closed, no?

Reply:
yea... this is the right solution to the double termination issue, not closing the signal handler

this was only necessary when the if condition only checked for Swept, since in that case the self.step(req, BackendState::Terminated) would cause a Signal::Interrupt before returning but after killing the backend, which got it back to the select! and ran the termination twice.